File: backend/internal/app/security.go
package app

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"os"
	"strings"
)

func hashPassword(password string) string {
	sum := sha256.Sum256([]byte("canal-plus:" + password))
	return hex.EncodeToString(sum[:])
}

func verifyPassword(password string, expectedHash string) bool {
	actual, err := hex.DecodeString(hashPassword(password))
	if err != nil {
		return false
	}
	expected, err := hex.DecodeString(expectedHash)
	if err != nil {
		return false
	}
	return hmac.Equal(actual, expected)
}

func secretKey() []byte {
	source := os.Getenv("CANAL_PLUS_SECRET")
	if source == "" {
		source = "canal-plus-dev-secret-change-me"
	}
	sum := sha256.Sum256([]byte(source))
	return sum[:]
}

func encryptText(plainText string) (string, error) {
	block, err := aes.NewCipher(secretKey())
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err := rand.Read(nonce); err != nil {
		return "", err
	}
	sealed := gcm.Seal(nil, nonce, []byte(plainText), nil)
	tagSize := gcm.Overhead()
	ciphertext := sealed[:len(sealed)-tagSize]
	tag := sealed[len(sealed)-tagSize:]
	encode := base64.RawURLEncoding.EncodeToString
	return "enc:v1:" + encode(nonce) + ":" + encode(tag) + ":" + encode(ciphertext), nil
}

func decryptText(secret string) (string, error) {
	if !strings.HasPrefix(secret, "enc:v1:") {
		return secret, nil
	}
	parts := strings.Split(secret, ":")
	if len(parts) != 5 {
		return "", fmt.Errorf("invalid encrypted value")
	}
	decode := base64.RawURLEncoding.DecodeString
	nonce, err := decode(parts[2])
	if err != nil {
		return "", err
	}
	tag, err := decode(parts[3])
	if err != nil {
		return "", err
	}
	ciphertext, err := decode(parts[4])
	if err != nil {
		return "", err
	}
	block, err := aes.NewCipher(secretKey())
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	sealed := append(ciphertext, tag...)
	plain, err := gcm.Open(nil, nonce, sealed, nil)
	if err != nil {
		return "", err
	}
	return string(plain), nil
}

func toPublicUser(user User) PublicUser {
	return PublicUser{
		ID:       user.ID,
		Name:     user.Name,
		Username: user.Username,
		Role:     user.Role,
	}
}

func toPublicDatasource(datasource Datasource) PublicDatasource {
	return PublicDatasource{
		ID:               datasource.ID,
		Name:             datasource.Name,
		Purpose:          datasource.Purpose,
		Host:             datasource.Host,
		Port:             datasource.Port,
		Username:         datasource.Username,
		DefaultSchema:    datasource.DefaultSchema,
		ConnectionStatus: datasource.ConnectionStatus,
		LastTestedAt:     datasource.LastTestedAt,
		LastTestMessage:  datasource.LastTestMessage,
		HasPassword:      datasource.PasswordSecret != "",
		IsDemo:           datasource.IsDemo,
		CreatedAt:        datasource.CreatedAt,
		UpdatedAt:        datasource.UpdatedAt,
	}
}

File: backend/src/security.ts
import { createCipheriv, createDecipheriv, createHash, randomBytes, timingSafeEqual } from "node:crypto";
import type { Datasource, PublicDatasource } from "./types.js";

function secretKey() {
  const source = process.env.CANAL_PLUS_SECRET || "canal-plus-dev-secret-change-me";
  return createHash("sha256").update(source).digest();
}

export function hashPassword(password: string) {
  return createHash("sha256").update(`canal-plus:${password}`).digest("hex");
}

export function verifyPassword(password: string, expectedHash: string) {
  const actual = Buffer.from(hashPassword(password), "hex");
  const expected = Buffer.from(expectedHash, "hex");
  return actual.length === expected.length && timingSafeEqual(actual, expected);
}

export function encryptText(plainText: string) {
  const iv = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", secretKey(), iv);
  const encrypted = Buffer.concat([cipher.update(plainText, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return ["enc:v1", iv.toString("base64url"), tag.toString("base64url"), encrypted.toString("base64url")].join(":");
}

export function decryptText(secret: string) {
  if (!secret.startsWith("enc:v1:")) {
    return secret;
  }

  const [, , ivText, tagText, encryptedText] = secret.split(":");
  const decipher = createDecipheriv("aes-256-gcm", secretKey(), Buffer.from(ivText, "base64url"));
  decipher.setAuthTag(Buffer.from(tagText, "base64url"));
  return Buffer.concat([
    decipher.update(Buffer.from(encryptedText, "base64url")),
    decipher.final()
  ]).toString("utf8");
}

export function toPublicDatasource(datasource: Datasource): PublicDatasource {
  const { passwordSecret, ...rest } = datasource;
  return {
    ...rest,
    hasPassword: Boolean(passwordSecret)
  };
}
